ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.computation;
+
+import io.confluent.ksql.engine.KsqlPlan;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.rest.entity.CommandId.Type;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Util for compacting the restore commands
+ */
+public final class RestoreCommandsCompactor {
+
+  private RestoreCommandsCompactor() {
+  }
+
+  /**
+   * Compact the list of commands to restore.
+   *
+   * <p>Finds any command's whose queries plans that are later terminated. Any such commands have
+   * their query plan and associated terminate command removed.
+   *
+   * <p>This compaction stops unnecessary creation of Streams topologies on a server restart.
+   * Building such topologies is relatively slow and best avoided.
+   *
+   * @param restoreCommands the list of commands to compact.
+   * @return the compacted list of commands.
+   */
+  static List<QueuedCommand> compact(final List<QueuedCommand> restoreCommands) {
```

Comment:
this feels needlessly complicated (in place modification using the iterator is really hard for me to follow as I need to track which source collection is being modified and used where). Can we simplify it by having a first pass just collect the index of terminated queries and the terminate commands in `restoreCommands` and then have a second pass that just copies the ones we need (modifying the ones that need to be modified) into a new list?

That would also obviate the need for using `IdentityHashMap` as we're using the index in the list to uniquely identify a command.

Reply:
Meh, I don't really see this as being complicated and, given this is throw away work as we're moving to key compacted topic, do we really want to spend more time on it?  You're basically asking me to totally re-write the algo.